private/pkg/storage/limit.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/atomic"
+)
+
+// LimitWriteBucket returns a [WriteBucket] that writes to [writeBucket]
+// but stops with an error after [limit] bytes are are written.
+//
+// The error can be checked using [IsWriteLimitReached].
+func LimitWriteBucket(writeBucket WriteBucket, limit int64) WriteBucket {
+	return newLimitedWriter(writeBucket, limit)
+}
+
+type limitedWriteBucket struct {
+	WriteBucket
+	currentSize *atomic.Int64
+	limit       int64
+}
+
+func newLimitedWriter(bucket WriteBucket, limit int64) *limitedWriteBucket {
```

Comment:
I would probably just make this an `int`, sort of a standard we have with interface types, and then correct <0 to be 0, ie `if limit < 0 { limit = 0 }`

Reply:
Updated the public function to accept an `int`